cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/logviewer/LogViewerWidget.kt

```diff
@@ -0,0 +1,126 @@
+import com.cognifide.cogboard.widget.connectionStrategy.ConnectionStrategy
+import com.cognifide.cogboard.widget.connectionStrategy.ConnectionStrategyFactory
+import com.cognifide.cogboard.widget.connectionStrategy.UnknownConnectionTypeException
+import com.cognifide.cogboard.widget.type.logviewer.logparser.LogParserStrategyFactory
+import io.vertx.core.Vertx
+import io.vertx.core.eventbus.MessageConsumer
+import io.vertx.core.json.JsonArray
+import io.vertx.core.json.JsonObject
+import main.kotlin.com.cognifide.cogboard.logStorage.LogStorageConfiguration
+import main.kotlin.com.cognifide.cogboard.logStorage.QuarantineRule
+
+class LogViewerWidget(
+    vertx: Vertx,
+    config: JsonObject,
+    serv: BoardsConfigService
+) : BaseWidget(vertx, config, serv) {
+    private val contentRepository: ContentRepository = ContentRepository.DEFAULT
+    private val address = config.endpointProp(Props.URL)
+    private var consumer: MessageConsumer<JsonObject>? = null
+    private val connectionStrategy: ConnectionStrategy? = determineConnectionStrategy()
+    private val logStorage: LogStorage? = connectionStrategy?.let {
+        LogStorage(
+                buildConfiguration(config),
+                it,
+                determineLogParsingStrategy()
+        )
+    }
+
+    init {
+        // Create a handler for updating the state of the widget.
+        createDynamicChangeSubscriber()?.handler { newState ->
+            newState?.body()?.let {
+                contentRepository.save(id, it)
+                logStorage?.rules = rules
+                logStorage?.filterExistingLogs()
+                updateWidget(false)
+            }
+        }
+    }
+
+    override fun start(): Widget {
+        vertx.deployVerticle(logStorage)
+        consumer = vertx.eventBus()
+            .consumer<JsonObject>(eventBusAddress)
+            .handler { logs ->
+                logs?.body()?.let { sendResponse(it) }
+            }
+        return super.start()
+    }
+
+    override fun stop(): Widget {
+        logStorage?.delete()
+        logStorage?.deploymentID()?.let { vertx.undeploy(it) }
+        consumer?.unregister()
+        return super.stop()
+    }
+
+    override fun updateState() {
+        updateWidget(true)
+    }
+
+    /** Updates the contents of the widget (optionally fetching new logs when [fetchNewLogs] is true). */
+    private fun updateWidget(fetchNewLogs: Boolean) {
+        if (address.isNotBlank()) {
+            logStorage?.rules = rules
+            logStorage?.updateLogs(fetchNewLogs)
+        } else {
+            sendConfigurationError("Endpoint URL is blank")
+        }
+    }
+
+    /** Sends the updated state to the client. */
+    private fun sendResponse(logs: JsonObject) {
+        val rules = contentRepository.get(id).getJsonArray(QUARANTINE_RULES) ?: JsonArray()
+        logs.put(QUARANTINE_RULES, rules)
+        send(logs)
+    }
+
+    /** Gets the quarantine rules from the */
```

Comment:
Fix: `... from the ?` either update or remove this comment

Reply:
Done.